### PR TITLE
Add python interface for topology ingestion

### PR DESCRIPTION
**Pull request by craffit**
### What does this PR do?

Add topology to the python interface. This explicitly does not include:
- having the StackStateAgentCheck in a separate repo
- having data published to stackstate

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/batcher/batcher_test.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+					Components:    []topology.Component{testComponent},
+					Relations:     []topology.Relation{},
+				},
+			},
+		})
+
+	batcher.Shutdown()
+}
+
+func TestBatchNoDataNoCommit(t *testing.T) {
```
Comment (TizianoPerrucci) [inline pkg/batcher/batcher_test.go]: `Complete` instead of `Commit`

File: pkg/batcher/batcher_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+					Components:    []topology.Component{},
+					Relations:     []topology.Relation{},
+				},
+			},
+		})
+
+	batcher.Shutdown()
+}
+
+func TestBatchFlushOnCommit(t *testing.T) {
```
Comment (TizianoPerrucci) [inline pkg/batcher/batcher_test.go]: `Complete` instead of `Commit`

File: pkg/batcher/topology_builder.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	// Amount of elements when we flush
+	flushElementCount int
+}
+
+// NewTopologyBuilder constructs a TopologyBuilder
+func NewTopologyBuilder(flushElementCount int) TopologyBuilder {
+	return TopologyBuilder{
+		topologies:        make(map[check.ID]topology.Topology),
+		elementCount:      0,
+		flushElementCount: flushElementCount,
```
Comment (TizianoPerrucci) [inline pkg/batcher/topology_builder.go]: I would rather name it to `maxCapacity`

File: pkg/batcher/batcher.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+// InitBatcher initializes the global batcher instance
+func InitBatcher(serializer serializer.AgentV1Serializer, hostname, agentName string, batchLimit int) {
+	batcherInit.Do(func() {
+		batcherInstance = newAsynchronousBatcher(serializer, hostname, agentName, batchLimit)
+	})
+}
+
+func newAsynchronousBatcher(serializer serializer.AgentV1Serializer, hostname, agentName string, batchLimit int) AsynchronousBatcher {
+	batcher := AsynchronousBatcher{
+		builder:    NewTopologyBuilder(batchLimit),
```
Comment (TizianoPerrucci) [inline pkg/batcher/batcher.go]: rename `batchLimit` according to the name we chosen in the TopologyBuilder

File: pkg/batcher/batcher.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	SubmitStartSnapshot(checkID check.ID, instance topology.Instance)
+	SubmitStopSnapshot(checkID check.ID, instance topology.Instance)
+	SubmitComplete(checkID check.ID)
+	Shutdown()
+}
+
+// InitBatcher initializes the global batcher instance
+func InitBatcher(serializer serializer.AgentV1Serializer, hostname, agentName string, batchLimit int) {
+	batcherInit.Do(func() {
+		batcherInstance = newAsynchronousBatcher(serializer, hostname, agentName, batchLimit)
```
Comment (TizianoPerrucci) [inline pkg/batcher/batcher.go]: rename `batchLimit` according to the name we chosen in the TopologyBuilder

File: pkg/batcher/topology_builder.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+import (
+	"github.com/StackVista/stackstate-agent/pkg/collector/check"
+	"github.com/StackVista/stackstate-agent/pkg/topology"
+)
+
+// Topologies is the type representing topologies gathered per check
+type Topologies map[check.ID]topology.Topology
+
+// TopologyBuilder is a helper class to build topology based on submitted data
```
Comment (TizianoPerrucci) [inline pkg/batcher/topology_builder.go]: I would add comments related to the fact that it is not thread-safe

File: pkg/config/config.go [hunk 7]
```diff
@@ -530,6 +530,15 @@ func GetMultipleEndpoints() (map[string][]string, error) {
 	return getMultipleEndpointsWithConfig(Datadog)
 }
 
+// GetBatcherLimit returns the mximum amount of elements per batcher for the batcher
+func GetBatcherLimit() int {
```
Comment (TizianoPerrucci) [inline pkg/config/config.go]: rename accordingly to name chosen in TopologyBuilder

File: pkg/collector/py/check.go [hunk 8]
```diff
@@ -80,6 +81,7 @@ func (c *PythonCheck) Run() error {
 	}
 
 	s.Commit()
+	batcher.GetBatcher().SubmitComplete(c.ID())
```
Comment (TizianoPerrucci) [inline pkg/collector/py/check.go]: if the check run produced errors or it is not able to retrieve the `sender` instance it fails; shouldn't we complete anyway if one of these things happens ?

File: pkg/collector/py/py.go [hunk 9]
```diff
@@ -118,7 +118,7 @@ func Initialize(paths ...string) *python.PyThreadState {
 	initKubeutil()     // `kubeutil` module if compiled in
 	initTagger()       // `tagger` module
 	initContainers()   // `containers` module
-
+	initStackState()   // The 'stackstate' module
```
[the remark this answers is not in the document]
Comment (craffit) [reply pkg/collector/py/py.go]: done

File: pkg/batcher/batcher_test.go (hunk 1 above)
Comment (craffit) [reply pkg/batcher/batcher_test.go]: done

File: pkg/batcher/batcher_test.go (hunk 2 above)
Comment (craffit) [reply pkg/batcher/batcher_test.go]: done

File: pkg/batcher/topology_builder.go (hunk 3 above)
Comment (craffit) [reply pkg/batcher/topology_builder.go]: done

File: pkg/batcher/batcher.go (hunk 4 above)
Comment (craffit) [reply pkg/batcher/batcher.go]: done

File: pkg/batcher/batcher.go (hunk 5 above)
Comment (craffit) [reply pkg/batcher/batcher.go]: done

File: pkg/batcher/topology_builder.go (hunk 6 above)
Comment (craffit) [reply pkg/batcher/topology_builder.go]: done

File: pkg/config/config.go (hunk 7 above)
Comment (craffit) [reply pkg/config/config.go]: what would you call the config item?

File: pkg/collector/py/check.go (hunk 8 above)
Comment (craffit) [reply pkg/collector/py/check.go]: agreed, i moved this to just after the run

**Pull request merged**